src/components/side-panel/interaction-panel/tooltip-config.js

```diff
@@ -18,40 +18,301 @@
+  } */
+`;
+
+const circleStyles = {
+  SHOW: {
+    stroke: '#fff',
+    fill: 'rgba(0,0,0,0)'
+  },
+  IDLE: {
+    stroke: '#A0A7B4',
+    fill: 'rgba(0,0,0,0)'
+  },
+  ACTIVE: {
+    stroke: '#1fbad6',
+    fill: '#1fbad6'
+  }
+};
+
+function getFormatLabels() {
+  return Object.values(TOOLTIP_FORMATS).map(v => v.label);
+}
+
+function getItemName(item) {
+  return item.name || item;
+}
+
 function TooltipConfigFactory(DatasetTag) {
-  const TooltipConfig = ({config, datasets, onChange}) => (
-    <div>
-      {Object.keys(config.fieldsToShow).map(dataId => (
-        <SidePanelSection key={dataId}>
-          <DatasetTag dataset={datasets[dataId]} />
-          <FieldSelector
-            fields={datasets[dataId].fields}
-            value={config.fieldsToShow[dataId]}
-            onSelect={fieldsToShow => {
-              const newConfig = {
-                ...config,
-                fieldsToShow: {
-                  ...config.fieldsToShow,
-                  [dataId]: fieldsToShow.map(d => d.name)
-                }
-              };
-              onChange(newConfig);
-            }}
-            closeOnSelect={false}
-            multiSelect
-            inputTheme="secondary"
-          />
-        </SidePanelSection>
-      ))}
-    </div>
-  );
+  class TooltipConfig extends Component {
+    getChickletAddon = dataId => {
+      return item => {
+        const {config, onChange} = this.props;
+        const [show, setShow] = React.useState(false);
+        const [selectionIndex, setSelectionIndex] = React.useState(0);
+        let circleStyle = selectionIndex ? circleStyles.ACTIVE : circleStyles.IDLE;
+        if (show) {
+          circleStyle = circleStyles.SHOW;
+        }
+        return (
+          <ChickletAddonWrapper>
+            <ChickletAddon data-tip data-for={`addon-${getItemName(item)}`}>
+              <Circle
+                height="8px"
+                onClick={e => {
+                  e.stopPropagation();
+                  setShow(Boolean(!show));
+                }}
+                {...circleStyle}
+              />
+              <Tooltip id={`addon-${getItemName(item)}`} effect="solid">
+                <span>
+                  {(selectionIndex && getFormatLabels()[selectionIndex]) || (
+                    <FormattedMessage id={'fieldSelector.formatting'} />
+                  )}
+                </span>
+              </Tooltip>
+            </ChickletAddon>
+            {show && (
+              <StyledPopover>
+                <DropdownList
+                  options={getFormatLabels()}
+                  selectionIndex={selectionIndex}
+                  onOptionSelected={(result, e) => {
+                    e.stopPropagation();
+                    setShow(false);
+                    setSelectionIndex(getFormatLabels().indexOf(result));
+                    const oldFieldsToShow = config.fieldsToShow[dataId];
+                    const fieldsToShow = oldFieldsToShow.map(field => {
```

Comment:
I don't like this logic handled inside the component. You are creating a hybrid `fieldsToShow` array, with a mix of `string` and `{name, format}`. in it. †his will create problem down the way, always have to handle both format.

The component should only accept `fieldsToShow` to be `{name, format}`. sometime `format == null`

To do that, you should

1. modify the `findFieldsToShow` in  `interaction-utils` so that it returns an array of objects
2. modify `InteractionSchemaV1` in `vis-state-schema.js` so that fieldsToShow is loaded from an array of `string` to `object` - I will give you more details on this later
3. modify `mergeInteractionTooltipConfig` in `vis-state-merger`, because now the `fieldsToShow` this function gets will be an array of object..

Reply:
@heshan0131 give me more details on #2

Reply:
[Resolved]